Tempfile V4

Stebalien opened the pull request:
This is a stack of changes for tempfile v4. My plan is to wait for the next Debian stable release then set the MSRV to that, although I may merge and cut an RC before then. This version tries to be minimally breaking in "normal" code but it will still break some code (ideally code that was already broken/buggy).

Other than the MSRV bump (letting us drop `once_cell`), changes include:

- Implement `AsRef<Path>` on `&NamedTempFile` (and friends) instead of directly on `NamedTempFile`. That way, functions accepting `AsRef<Path>` won't be able to take (then drop) `NamedTempFile` by value. See https://github.com/Stebalien/tempfile/issues/115.
- Temporary directories are created with 0700 permissions by default instead of 0755 to match temporary files.
- Temporary filenames no longer start with a `.` by default. Temporary files shouldn't generally be hidden. Additionally, it's now possible to configure the default temporary-filename prefix via the `env` package.
- The `Builder` can now has a single lifetime instead of separate ones for `prefix` and `suffix`. It can also be constructed with a const `Builder::new` if desired.
- Temporary file names are validated to be valid file-*names*, not including path separators, etc. It's still not safe to allow an attacker to control the prefix/suffix, but this gives us a little extra safety.
- The closure passed to `Builder::make` used to take a `&Path` but now takes an `&MakeParams` struct (that dereferences to a `&Path`). This lets us pass additional builder parameters like the desired permissions.
- The deprecated `TempDir::into_path` is removed in favor of `TempDir::keep`.
- `SpooledTempDir::new` is now const.
- `tempfile::env::temp_dir()` now returns a result and won't panic on platforms where `std::env::temp_dir` would panic.
- Unnamed temporary files may now be created in inaccessible directories (directories where we're missing search permissions on one of the parent directories).

TBD:

- [ ] Prevent temporary file cleanup by taking a shared advisory locks on all named temporary files to prevent temporary file cleaners from deleting them. This could break some things so we have to be careful, but taking a shared lock is probably reasonable? Possibly only do this if the parent directory is `+t` (sticky)? We can also make it configurable in the builder, the question will be: what's the default behavior?

Fixes:

- #59 
- Everything in #258
- #40

Stebalien committed:
fix: implement AsRef by-reference

Otherwise it's really easy to pass a temporary file handle by reference,
causing it to be dropped by the receiver deleting the underlying file.

fixes #115

Stebalien committed:
feat: forward Seek::rewind

Stebalien committed:
fix: create temporary directories with 0700 permissions by default

fixes #284

Stebalien committed:
feat: misc builder improvements and better defaults

1. Default to "tmp" as the file prefix instead of ".tmp".
2. Make it possible to construct the builder in a constfn.
3. Use a single lifetime in the builder.

Stebalien committed:
fix: forbid path separators in prefix/suffix

fixes #59

Stebalien committed:
feat: change Builder::make to take a parameters struct

That way, we can pass through additional builder options. This new
struct implements AsRef<Path> and Deref<Target=Path> so breakage should
be minimal.

Stebalien committed:
feat: update env

Stebalien committed:
feat: drop once_cell dependency

Stebalien committed:
feat: use new absolute path function

Stebalien committed:
feat: update MSRV from 1.63 to 1.85

Imberflur commented:
One minor annoyance I noticed with `TempDir::close` is that you no longer have access to the path for error reporting, so it has to be copied beforehand.

Stebalien committed:
feat: remove deprecated methods

Stebalien committed:
feat: allow const spooled temporary files

Cursors are stable in rust as of 1.79. This change lets us keep a global
scratch file behind a Mutex.

Stebalien committed:
fix: avoid panicing on platforms with no sane default tempdir

fixes #365

NobodyXu commented:
debian trixie has released with rust 1.85, though IIRC one of the other popular distro (centos or fedora, I can't remember) still on 1.84

Stebalien commented:
> though IIRC one of the other popular distro (centos or fedora, I can't remember) still on 1.84

Ah, that's good to know. I'll see if I can downgrade to that.

Also note, I'm still planing on maintaining v3 for the foreseeable future.

Stebalien commented:
It's Ubuntu plucky. Which is unfortunately rules out 2024 edition unless I wait until October (although it might take that long anyways if I try to tackle the remaining TODOs).

Stebalien committed:
feat: switch to rust 2024 edition

Stebalien committed:
chore: fix clippy lints

Stebalien commented:
This now bumps the MSRV to 1.85 and the edition to 2024.

Stebalien committed:
ci: use the current compiler version to resolve dependencies

I'm testing specific compiler versions, not the MSRV. This seems to be
the easiest way to do this reliably.

I've also changed the cargo config to by-default allow incompatible
versions. That way, people running `cargo test` locally will get the
latest dependencies instead of outdated dependencies. I expect anyone
working on tempfile to use the latest compiler anyways.

Also, auto-detect MSRV

Stebalien committed:
feat: only require absolute paths for named temporary files

For unnamed temporary files, we now handle the creation/deletion race by
opening the target directory, creating our file relative to that
directory, then deleting the file relative to that directory.

fixes #40

Stebalien commented:
This is now blocked on:

1. A libc release for `openat` on RedoxOS
2. https://github.com/bytecodealliance/rustix/pull/1594 (blocking updating libc in rustix).
3. A PR in rustix to implement `openat` and `unlinkat` for RedoxOS.
4. A rustix release.
